File: zig/src/rename.zig
const std = @import("std");
const types = @import("types.zig");
const util = @import("util.zig");
const txn = @import("txn.zig");
const block_mod = @import("block.zig");

pub fn doRename(self: anytype, old_lastname: []const u8, old_blockindex: u32, old_type_dir: u8, new_lastname: []const u8, new_blockindex: u32, new_type_dir: u8) i32 {
    var b4: [4]u8 = undefined;
    var b2: [2]u8 = undefined;
    var s: [types.BlockNameMaxSize + 2]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 2);

    var old_ba: [4]types.BlockArray = .{ .{}, .{}, .{}, .{} };
    var old_ba_used: usize = 0;
    var old_block_head: []u8 = undefined;
    var old_block_last: []u8 = undefined;
    var old_block_item: []u8 = undefined;
    var old_block_prev: []u8 = undefined;
    var old_block_item_index: u32 = 0;
    var old_block_last_index: u32 = 0;
    var old_block_prev_index: u32 = 0;
    var old_block_head_index: u32 = 0;

    var old_block = util.blockZero();
    if (!block_mod.readblock(self, old_blockindex, old_block[0..])) return 1;
    @memcpy(old_ba[0].block[0..], old_block[0..]);
    old_ba[0].blockindex = old_blockindex;
    old_ba[0].active = true;
    old_block_head = old_ba[0].block[0..];
    old_block_head_index = old_blockindex;
    old_ba_used = 1;

    @memcpy(b4[0..], old_block_head[12 + 1 + 14 + 4 .. 12 + 1 + 14 + 8]);
    const old_stop_blockindex = util.b4ToU32(b4[0..]);
    @memcpy(b2[0..], old_block_head[12 + 1 + 14 + 4 + 4 .. 12 + 1 + 14 + 4 + 6]);
    var old_offset = util.b2ToU16(b2[0..]);

    if (old_stop_blockindex == old_block_head_index) {
        old_block_last = old_block_head;
        old_block_last_index = old_block_head_index;
    } else {
        if (!block_mod.readblock(self, old_stop_blockindex, old_ba[1].block[0..])) return 1;
        old_ba[1].blockindex = old_stop_blockindex;
        old_ba[1].active = true;
        old_block_last = old_ba[1].block[0..];
        old_block_last_index = old_stop_blockindex;
        old_ba_used += 1;
    }

    var old_item_offset: u16 = 0;
    var old_dir_file: u8 = 0;
    var flag = false;
    var index = old_block_head_index;
    while (true) {
        var k: usize = types.BlockHead;
        var item: usize = 0;
        while (item < types.BlockItemMaxCount) : (item += 1) {
            const state = old_block[k];
            k += 1;
            @memset(s[0..], 0);
            @memcpy(s[0..types.BlockNameMaxSize], old_block[k .. k + types.BlockNameMaxSize]);
            k += types.BlockNameMaxSize;
            if (!std.mem.eql(u8, util.fixedCStr(s[0..types.BlockNameMaxSize + 1]), old_lastname)) {
                k += 10;
                if (index == old_stop_blockindex and k + 1 >= util.u16ToUsize(old_offset)) return 4;
                continue;
            }
            old_dir_file = state & 0x01;
            if (old_type_dir == 1 and old_dir_file == 1) return 2;
            if (new_type_dir == 1 and old_dir_file == 1) return 6;
            old_item_offset = util.usizeToU16(k + 10);
            var u = false;
            var j: usize = 0;
            while (j < old_ba_used) : (j += 1) {
                if (old_ba[j].blockindex == index) {
                    old_block_item = old_ba[j].block[0..];
                    old_block_item_index = index;
                    u = true;
                    break;
                }
            }
            if (!u) {
                @memcpy(old_ba[old_ba_used].block[0..], old_block[0..]);
                old_ba[old_ba_used].blockindex = index;
                old_ba[old_ba_used].active = true;
                old_block_item = old_ba[old_ba_used].block[0..];
                old_block_item_index = index;
                old_ba_used += 1;
            }
            flag = true;
            break;
        }
        if (flag) break;
        @memcpy(b4[0..], old_block[4..8]);
        index = util.b4ToU32(b4[0..]);
        if (index == 0) return 1;
        if (!block_mod.readblock(self, index, old_block[0..])) return 1;
    }

    var new_ba: [2]types.BlockArray = .{ .{}, .{} };
    var new_ba_used: usize = 0;
    var new_block_head: []u8 = undefined;
    var new_block_last: []u8 = undefined;
    var new_block_head_index: u32 = 0;
    var new_block_last_index: u32 = 0;

    var new_block = util.blockZero();
    if (!block_mod.readblock(self, new_blockindex, new_block[0..])) return 1;
    @memcpy(new_ba[0].block[0..], new_block[0..]);
    new_ba[0].blockindex = new_blockindex;
    new_ba[0].active = true;
    new_block_head = new_ba[0].block[0..];
    new_block_head_index = new_blockindex;
    new_ba_used = 1;

    @memcpy(b4[0..], new_block_head[12 + 1 + 14 + 4 .. 12 + 1 + 14 + 8]);
    const new_stop_blockindex = util.b4ToU32(b4[0..]);
    @memcpy(b2[0..], new_block_head[12 + 1 + 14 + 4 + 4 .. 12 + 1 + 14 + 4 + 6]);
    var new_offset = util.b2ToU16(b2[0..]);

    if (new_stop_blockindex == new_block_head_index) {
        new_block_last = new_block_head;
        new_block_last_index = new_block_head_index;
    } else {
        if (!block_mod.readblock(self, new_stop_blockindex, new_ba[1].block[0..])) return 1;
        new_ba[1].blockindex = new_stop_blockindex;
        new_ba[1].active = true;
        new_block_last = new_ba[1].block[0..];
        new_block_last_index = new_stop_blockindex;
        new_ba_used += 1;
    }

    flag = false;
    index = new_block_head_index;
    while (true) {
        var k: usize = types.BlockHead;
        var item: usize = 0;
        while (item < types.BlockItemMaxCount) : (item += 1) {
            k += 1;
            @memset(s[0..], 0);
            @memcpy(s[0..types.BlockNameMaxSize], new_block[k .. k + types.BlockNameMaxSize]);
            k += types.BlockNameMaxSize;
            if (!std.mem.eql(u8, util.fixedCStr(s[0..types.BlockNameMaxSize + 1]), new_lastname)) {
                k += 10;
                if (index == new_stop_blockindex and k + 1 >= util.u16ToUsize(new_offset)) {
                    flag = true;
                    break;
                }
                continue;
            }
            return 5;
        }
        if (flag) break;
        @memcpy(b4[0..], new_block[4..8]);
        index = util.b4ToU32(b4[0..]);
        if (index == 0) return 1;
        if (!block_mod.readblock(self, index, new_block[0..])) return 1;
    }

    if (old_block_head_index == new_block_head_index) {
        util.copyNameInto(old_block_item[old_item_offset - 24 .. old_item_offset - 10], new_lastname);
        if (self.tmp.state == 0 and !txn.tmpstart(self, 1)) return 1;
        if (!block_mod.writeblock(self, old_block_item_index, old_block_item)) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        if (self.tmp.state == 1 and !txn.commit(self)) return 1;
        return 0;
    }

    if (self.tmp.state == 0 and !txn.tmpstart(self, 1)) return 1;

    if (old_dir_file == 0) {
        @memcpy(b4[0..], old_block_item[old_item_offset - 10 .. old_item_offset - 6]);
        const path_blockindex = util.b4ToU32(b4[0..]);
        var path_block = util.blockZero();
        if (!block_mod.readblock(self, path_blockindex, path_block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        util.u32ToB4(new_block_head_index, &b4);
        @memcpy(path_block[types.BlockHead + 25 + 1 + 14 .. types.BlockHead + 25 + 1 + 18], b4[0..]);
        if (!block_mod.writeblock(self, path_blockindex, path_block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
    }

    var block2 = util.blockZero();
    var blockindex2: u32 = 0;
    if (util.u16ToUsize(new_offset) < types.BlockSize) {
        std.mem.copyForwards(u8, new_block_last[new_offset .. new_offset + 25], old_block_item[old_item_offset - 25 .. old_item_offset]);
        new_offset += 25;
        util.u16ToB2(new_offset, &b2);
        @memcpy(new_block_head[types.BlockOffset .. types.BlockOffset + 2], b2[0..]);
    } else {
        blockindex2 = block_mod.genblockindex(self);
        if (blockindex2 == 0) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        util.u32ToB4(new_block_last_index, &b4);
        @memcpy(block2[8..12], b4[0..]);
        std.mem.copyForwards(u8, block2[types.BlockHead .. types.BlockHead + 25], old_block_item[old_item_offset - 25 .. old_item_offset]);
        if (!block_mod.writeblock(self, blockindex2, block2[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        util.u32ToB4(blockindex2, &b4);
        @memcpy(new_block_last[4..8], b4[0..]);
        new_offset = types.BlockHead + 25;
        util.u16ToB2(new_offset, &b2);
        @memcpy(new_block_head[types.BlockOffset .. types.BlockOffset + 2], b2[0..]);
        @memcpy(new_block_head[types.BlockStopBlockindex .. types.BlockStopBlockindex + 4], b4[0..]);
    }
    for (new_ba[0..new_ba_used]) |entry| {
        if (!entry.active) continue;
        if (!block_mod.writeblock(self, entry.blockindex, entry.block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
    }

    if (old_block_item_index != old_stop_blockindex or old_item_offset != old_offset) {
        std.mem.copyForwards(u8, old_block_item[old_item_offset - 25 .. old_item_offset], old_block_last[old_offset - 25 .. old_offset]);
    }
    old_offset -= 25;
    util.u16ToB2(old_offset, &b2);
    @memcpy(old_block_head[types.BlockOffset .. types.BlockOffset + 2], b2[0..]);

    if (old_offset < 25) {
        @memcpy(b4[0..], old_block_last[8..12]);
        old_block_prev_index = util.b4ToU32(b4[0..]);
        _ = block_mod.removeblock(self, old_block_last_index);
        var k_idx: isize = -1;
        var i: usize = 0;
        while (i < old_ba_used) : (i += 1) {
            if (old_ba[i].blockindex == old_block_last_index) {
                old_ba[i].active = false;
                k_idx = @intCast(i);
                break;
            }
        }
        if (k_idx < 0) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
        var u = false;
        i = 0;
        while (i < old_ba_used) : (i += 1) {
            if (old_ba[i].blockindex == old_block_prev_index) {
                old_block_prev = old_ba[i].block[0..];
                u = true;
                break;
            }
        }
        if (!u) {
            if (!block_mod.readblock(self, old_block_prev_index, old_block[0..])) {
                if (self.tmp.state == 1) txn.tmpstop(self);
                return 1;
            }
            @memcpy(old_ba[@intCast(k_idx)].block[0..], old_block[0..]);
            old_ba[@intCast(k_idx)].blockindex = old_block_prev_index;
            old_ba[@intCast(k_idx)].active = true;
            old_block_prev = old_ba[@intCast(k_idx)].block[0..];
        }
        @memset(old_block_prev[4..8], 0);
        util.u32ToB4(old_block_prev_index, &b4);
        @memcpy(old_block_head[types.BlockStopBlockindex .. types.BlockStopBlockindex + 4], b4[0..]);
        old_offset = types.BlockSize;
        util.u16ToB2(old_offset, &b2);
        @memcpy(old_block_head[types.BlockOffset .. types.BlockOffset + 2], b2[0..]);
    }
    for (old_ba[0..old_ba_used]) |entry| {
        if (!entry.active) continue;
        if (!block_mod.writeblock(self, entry.blockindex, entry.block[0..])) {
            if (self.tmp.state == 1) txn.tmpstop(self);
            return 1;
        }
    }
    if (self.tmp.state == 1 and !txn.commit(self)) return 1;
    return 0;
}

pub fn rename(self: anytype, old_name: []const u8, new_name: []const u8) i32 {
    if (self.fp == null or old_name.len == 0 or new_name.len == 0) return 1;

    var blockindex: u32 = undefined;
    var start: usize = undefined;
    if (old_name[0] == '/') {
        blockindex = 1;
        start = 1;
    } else {
        blockindex = util.currentPwdBlockindex(self);
        start = 0;
    }
    var s: [types.BlockNameMaxSize + 2]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 2);
    var slen: usize = 0;
    var i: usize = start;
    while (i < old_name.len) : (i += 1) {
        if (old_name[i] == '/') {
            if (slen == 0) continue;
            if (i == old_name.len - 1) break;
            const index = block_mod.findPathBlockindex(self, blockindex, s[0..slen]);
            if (index < 1) return 2;
            blockindex = index;
            slen = 0;
            continue;
        }
        s[slen] = old_name[i];
        slen += 1;
        if (slen > types.BlockNameMaxSize) return 2;
    }
    var old_lastname_buf: [types.BlockNameMaxSize + 1]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 1);
    @memcpy(old_lastname_buf[0..slen], s[0..slen]);
    const old_lastname = old_lastname_buf[0..slen];
    if (std.mem.eql(u8, old_lastname, ".") or std.mem.eql(u8, old_lastname, "..")) return 2;
    const old_blockindex = blockindex;
    const old_type_dir: u8 = if (old_name[old_name.len - 1] == '/') 1 else 0;

    if (new_name[0] == '/') {
        blockindex = 1;
        start = 1;
    } else {
        blockindex = util.currentPwdBlockindex(self);
        start = 0;
    }
    slen = 0;
    i = start;
    while (i < new_name.len) : (i += 1) {
        if (new_name[i] == '/') {
            if (slen == 0) continue;
            if (i == new_name.len - 1) break;
            const index = block_mod.findPathBlockindex(self, blockindex, s[0..slen]);
            if (index < 1) return 3;
            blockindex = index;
            slen = 0;
            continue;
        }
        s[slen] = new_name[i];
        slen += 1;
        if (slen > types.BlockNameMaxSize) return 3;
    }
    var new_lastname_buf: [types.BlockNameMaxSize + 1]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 1);
    @memcpy(new_lastname_buf[0..slen], s[0..slen]);
    const new_lastname = new_lastname_buf[0..slen];
    if (std.mem.eql(u8, new_lastname, ".") or std.mem.eql(u8, new_lastname, "..")) return 3;
    const new_blockindex = blockindex;
    const new_type_dir: u8 = if (new_name[new_name.len - 1] == '/') 1 else 0;

    return doRename(self, old_lastname, old_blockindex, old_type_dir, new_lastname, new_blockindex, new_type_dir);
}

File: cpp/src/rename.cpp
#include "types.hpp"

#include "util.hpp"

#include <algorithm>

namespace {

int do_rename(filefs::detail::FileSystemState& ffs, const std::string_view old_lastname,
              const std::uint32_t old_blockindex, const std::uint8_t old_type_dir,
              const std::string_view new_lastname, const std::uint32_t new_blockindex,
              const std::uint8_t new_type_dir) {
  using namespace filefs;

  std::array<detail::BlockArray, 4> old_ba{};
  int old_ba_used = 0;
  detail::Block* old_block_head = nullptr;
  detail::Block* old_block_last = nullptr;
  detail::Block* old_block_item = nullptr;
  detail::Block* old_block_prev = nullptr;
  std::uint32_t old_block_item_index = 0U;
  std::uint32_t old_block_last_index = 0U;
  std::uint32_t old_block_prev_index = 0U;
  std::uint32_t old_block_head_index = 0U;

  detail::Block old_block{};
  if (!ffs.readblock(old_blockindex, old_block)) {
    return 1;
  }
  old_ba[0].block = old_block;
  old_ba[0].blockindex = old_blockindex;
  old_ba[0].active = 1U;
  old_block_head = &old_ba[0].block;
  old_block_head_index = old_blockindex;
  old_ba_used = 1;

  const std::uint32_t old_stop_blockindex =
      detail::b4_to_u32(std::span<const std::uint8_t, 4>(old_block_head->data() + 12 + 1 + 14 + 4, 4));
  std::uint16_t old_offset =
      detail::b2_to_u16(std::span<const std::uint8_t, 2>(old_block_head->data() + 12 + 1 + 14 + 4 + 4, 2));

  if (old_stop_blockindex == old_block_head_index) {
    old_block_last = old_block_head;
    old_block_last_index = old_block_head_index;
  } else {
    if (!ffs.readblock(old_stop_blockindex, old_ba[1].block)) {
      return 1;
    }
    old_ba[1].blockindex = old_stop_blockindex;
    old_ba[1].active = 1U;
    old_block_last = &old_ba[1].block;
    old_block_last_index = old_stop_blockindex;
    ++old_ba_used;
  }

  std::uint16_t old_item_offset = 0U;
  std::uint8_t old_dir_file = 0U;
  bool flag = false;
  std::uint32_t index = old_block_head_index;
  for (;;) {
    int k = detail::BlockHead;
    for (int i = 0; i < detail::BlockItemMaxCount; ++i) {
      const std::uint8_t state = old_block[static_cast<std::size_t>(k)];
      ++k;
      const bool name_match =
          detail::fixed_cstr(std::span<const std::uint8_t>(old_block.data() + k, BlockNameMaxSize)) ==
          old_lastname;
      k += static_cast<int>(BlockNameMaxSize);
      if (!name_match) {
        k += 10;
        if (index == old_stop_blockindex && k + 1 >= static_cast<int>(old_offset)) {
          return 4;
        }
        continue;
      }
      old_dir_file = state & 0x01U;
      if (old_type_dir == 1U && old_dir_file == 1U) {
        return 2;
      }
      if (new_type_dir == 1U && old_dir_file == 1U) {
        return 6;
      }
      old_item_offset = static_cast<std::uint16_t>(k + 10);

      bool reused = false;
      for (int j = 0; j < old_ba_used; ++j) {
        if (old_ba[j].blockindex == index) {
          old_block_item = &old_ba[j].block;
          old_block_item_index = index;
          reused = true;
          break;
        }
      }
      if (!reused) {
        old_ba[old_ba_used].block = old_block;
        old_ba[old_ba_used].blockindex = index;
        old_ba[old_ba_used].active = 1U;
        old_block_item = &old_ba[old_ba_used].block;
        old_block_item_index = index;
        ++old_ba_used;
      }
      flag = true;
      break;
    }
    if (flag) {
      break;
    }
    index = detail::b4_to_u32(std::span<const std::uint8_t, 4>(old_block.data() + 4, 4));
    if (index == 0U || !ffs.readblock(index, old_block)) {
      return 1;
    }
  }

  std::array<detail::BlockArray, 2> new_ba{};
  int new_ba_used = 0;
  detail::Block* new_block_head = nullptr;
  detail::Block* new_block_last = nullptr;
  std::uint32_t new_block_head_index = 0U;
  std::uint32_t new_block_last_index = 0U;

  detail::Block new_block{};
  if (!ffs.readblock(new_blockindex, new_block)) {
    return 1;
  }
  new_ba[0].block = new_block;
  new_ba[0].blockindex = new_blockindex;
  new_ba[0].active = 1U;
  new_block_head = &new_ba[0].block;
  new_block_head_index = new_blockindex;
  new_ba_used = 1;

  const std::uint32_t new_stop_blockindex =
      detail::b4_to_u32(std::span<const std::uint8_t, 4>(new_block_head->data() + 12 + 1 + 14 + 4, 4));
  std::uint16_t new_offset =
      detail::b2_to_u16(std::span<const std::uint8_t, 2>(new_block_head->data() + 12 + 1 + 14 + 4 + 4, 2));

  if (new_stop_blockindex == new_block_head_index) {
    new_block_last = new_block_head;
    new_block_last_index = new_block_head_index;
  } else {
    if (!ffs.readblock(new_stop_blockindex, new_ba[1].block)) {
      return 1;
    }
    new_ba[1].blockindex = new_stop_blockindex;
    new_ba[1].active = 1U;
    new_block_last = &new_ba[1].block;
    new_block_last_index = new_stop_blockindex;
    ++new_ba_used;
  }

  flag = false;
  index = new_block_head_index;
  for (;;) {
    int k = detail::BlockHead;
    for (int i = 0; i < detail::BlockItemMaxCount; ++i) {
      ++k;
      const bool name_match =
          detail::fixed_cstr(std::span<const std::uint8_t>(new_block.data() + k, BlockNameMaxSize)) ==
          new_lastname;
      k += static_cast<int>(BlockNameMaxSize);
      if (!name_match) {
        k += 10;
        if (index == new_stop_blockindex && k + 1 >= static_cast<int>(new_offset)) {
          flag = true;
          break;
        }
        continue;
      }
      return 5;
    }
    if (flag) {
      break;
    }
    index = detail::b4_to_u32(std::span<const std::uint8_t, 4>(new_block.data() + 4, 4));
    if (index == 0U || !ffs.readblock(index, new_block)) {
      return 1;
    }
  }

  if (old_block_head_index == new_block_head_index) {
    std::fill_n(old_block_item->data() + old_item_offset - 10 - 14, BlockNameMaxSize, 0U);
    std::memcpy(old_block_item->data() + old_item_offset - 10 - 14, new_lastname.data(), new_lastname.size());
    if (ffs.tmp.state == 0U && !ffs.tmpstart(1U)) {
      return 1;
    }
    if (!ffs.writeblock(old_block_item_index, *old_block_item)) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
    if (ffs.tmp.state == 1U && !ffs.commit()) {
      return 1;
    }
    return 0;
  }

  if (ffs.tmp.state == 0U && !ffs.tmpstart(1U)) {
    return 1;
  }

  if (old_dir_file == 0U) {
    const std::uint32_t path_blockindex =
        detail::b4_to_u32(std::span<const std::uint8_t, 4>(old_block_item->data() + old_item_offset - 10, 4));
    detail::Block path_block{};
    if (!ffs.readblock(path_blockindex, path_block)) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
    detail::u32_to_b4(new_block_head_index,
                      std::span<std::uint8_t, 4>(path_block.data() + detail::BlockHead + 25 + 1 + 14, 4));
    if (!ffs.writeblock(path_blockindex, path_block)) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
  }

  detail::Block block2{};
  std::uint32_t blockindex2 = 0U;
  if (new_offset < BlockSize) {
    std::copy_n(old_block_item->data() + old_item_offset - 25, 25, new_block_last->data() + new_offset);
    new_offset = static_cast<std::uint16_t>(new_offset + 25);
    std::array<std::uint8_t, 2> b2{};
    detail::u16_to_b2(new_offset, b2);
    std::copy(b2.begin(), b2.end(), new_block_head->begin() + detail::BlockOffset);
  } else {
    blockindex2 = ffs.genblockindex();
    if (blockindex2 == 0U) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
    block2.fill(0U);
    detail::u32_to_b4(new_block_last_index, std::span<std::uint8_t, 4>(block2.data() + 8, 4));
    std::copy_n(old_block_item->data() + old_item_offset - 25, 25, block2.data() + detail::BlockHead);
    if (!ffs.writeblock(blockindex2, block2)) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
    detail::u32_to_b4(blockindex2, std::span<std::uint8_t, 4>(new_block_last->data() + 4, 4));
    new_offset = static_cast<std::uint16_t>(detail::BlockHead + 25);
    std::array<std::uint8_t, 2> b2{};
    detail::u16_to_b2(new_offset, b2);
    std::copy(b2.begin(), b2.end(), new_block_head->begin() + detail::BlockOffset);
    detail::u32_to_b4(blockindex2, std::span<std::uint8_t, 4>(new_block_head->data() + detail::BlockStopBlockindex, 4));
  }
  for (int i = 0; i < new_ba_used; ++i) {
    if (new_ba[i].active != 0U && !ffs.writeblock(new_ba[i].blockindex, new_ba[i].block)) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
  }

  if (old_block_item_index != old_stop_blockindex || old_item_offset != old_offset) {
    std::copy_n(old_block_last->data() + old_offset - 25, 25, old_block_item->data() + old_item_offset - 25);
  }
  old_offset = static_cast<std::uint16_t>(old_offset - 25);
  std::array<std::uint8_t, 2> b2{};
  detail::u16_to_b2(old_offset, b2);
  std::copy(b2.begin(), b2.end(), old_block_head->begin() + detail::BlockOffset);

  if (old_offset < 25U) {
    old_block_prev_index = detail::b4_to_u32(std::span<const std::uint8_t, 4>(old_block_last->data() + 8, 4));
    if (!ffs.removeblock(old_block_last_index)) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
    int slot = -1;
    for (int i = 0; i < old_ba_used; ++i) {
      if (old_ba[i].blockindex == old_block_last_index) {
        old_ba[i].active = 0U;
        slot = i;
        break;
      }
    }
    if (slot < 0) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
    bool reused = false;
    for (int i = 0; i < old_ba_used; ++i) {
      if (old_ba[i].blockindex == old_block_prev_index) {
        old_block_prev = &old_ba[i].block;
        reused = true;
        break;
      }
    }
    if (!reused) {
      if (!ffs.readblock(old_block_prev_index, old_block)) {
        if (ffs.tmp.state == 1U) {
          ffs.tmpstop();
        }
        return 1;
      }
      old_ba[slot].block = old_block;
      old_ba[slot].blockindex = old_block_prev_index;
      old_ba[slot].active = 1U;
      old_block_prev = &old_ba[slot].block;
    }
    std::fill_n(old_block_prev->data() + 4, 4, 0U);
    detail::u32_to_b4(old_block_prev_index,
                      std::span<std::uint8_t, 4>(old_block_head->data() + detail::BlockStopBlockindex, 4));
    old_offset = BlockSize;
    detail::u16_to_b2(old_offset, b2);
    std::copy(b2.begin(), b2.end(), old_block_head->begin() + detail::BlockOffset);
  }

  for (int i = 0; i < old_ba_used; ++i) {
    if (old_ba[i].active != 0U && !ffs.writeblock(old_ba[i].blockindex, old_ba[i].block)) {
      if (ffs.tmp.state == 1U) {
        ffs.tmpstop();
      }
      return 1;
    }
  }
  if (ffs.tmp.state == 1U && !ffs.commit()) {
    return 1;
  }
  return 0;
}

}  // namespace

namespace filefs {

int FileSystem::rename(const std::string_view old_name, const std::string_view new_name) {
  if (!state_ || !state_->mounted() || old_name.empty() || new_name.empty()) {
    return 1;
  }

  int len_n = static_cast<int>(old_name.size());
  std::uint32_t blockindex = 0U;
  int start = 0;
  if (old_name.front() == '/') {
    blockindex = 1U;
    start = 1;
  } else if (state_->tmp.state == 0U) {
    blockindex = state_->pwd_blockindex;
  } else {
    blockindex = state_->tmp.pwd_blockindex;
  }

  std::array<char, BlockNameMaxSize + 2> s{};
  int slen = 0;
  for (int i = start; i < len_n; ++i) {
    if (old_name[static_cast<std::size_t>(i)] == '/') {
      if (slen == 0) {
        continue;
      }
      if (i == len_n - 1) {
        break;
      }
      const std::uint32_t index =
          state_->find_path_blockindex(blockindex, std::string_view{s.data(), static_cast<std::size_t>(slen)});
      if (index < 1U) {
        return 2;
      }
      blockindex = index;
      slen = 0;
      continue;
    }
    s[static_cast<std::size_t>(slen++)] = old_name[static_cast<std::size_t>(i)];
    if (slen > static_cast<int>(BlockNameMaxSize)) {
      return 2;
    }
  }
  if (slen > static_cast<int>(BlockNameMaxSize)) {
    return 2;
  }
  const std::string old_lastname{s.data(), static_cast<std::size_t>(slen)};
  if (old_lastname == "." || old_lastname == "..") {
    return 2;
  }
  const std::uint32_t old_blockindex = blockindex;
  std::uint8_t old_type_dir = old_name.back() == '/' ? 1U : 0U;

  len_n = static_cast<int>(new_name.size());
  if (new_name.front() == '/') {
    blockindex = 1U;
    start = 1;
  } else if (state_->tmp.state == 0U) {
    blockindex = state_->pwd_blockindex;
  } else {
    blockindex = state_->tmp.pwd_blockindex;
  }
  slen = 0;
  for (int i = start; i < len_n; ++i) {
    if (new_name[static_cast<std::size_t>(i)] == '/') {
      if (slen == 0) {
        continue;
      }
      if (i == len_n - 1) {
        break;
      }
      const std::uint32_t index =
          state_->find_path_blockindex(blockindex, std::string_view{s.data(), static_cast<std::size_t>(slen)});
      if (index < 1U) {
        return 3;
      }
      blockindex = index;
      slen = 0;
      continue;
    }
    s[static_cast<std::size_t>(slen++)] = new_name[static_cast<std::size_t>(i)];
    if (slen > static_cast<int>(BlockNameMaxSize)) {
      return 3;
    }
  }
  if (slen > static_cast<int>(BlockNameMaxSize)) {
    return 3;
  }
  const std::string new_lastname{s.data(), static_cast<std::size_t>(slen)};
  if (new_lastname == "." || new_lastname == "..") {
    return 3;
  }
  const std::uint32_t new_blockindex = blockindex;
  std::uint8_t new_type_dir = new_name.back() == '/' ? 1U : 0U;

  return do_rename(*state_, old_lastname, old_blockindex, old_type_dir, new_lastname, new_blockindex, new_type_dir);
}

}  // namespace filefs
